File: 242.有效的字母异位词.py
#
# @lc app=leetcode.cn id=242 lang=python3
#
# [242] 有效的字母异位词
#

# @lc code=start
class Solution:
    def isAnagram(self, s: str, t: str) -> bool:
        letter_dict = {}
        index = 0
        while index < min(len(s), len(t)):
            if s[index] not in letter_dict:
                letter_dict[s[index]] = 1
            else:
                letter_dict[s[index]] += 1
            
            if t[index] not in letter_dict:
                letter_dict[t[index]] = -1
            else:
                letter_dict[t[index]] -= 1
            index += 1
        while index < len(s):
            if s[index] not in letter_dict:
                letter_dict[s[index]] = 1
            else:
                letter_dict[s[index]] += 1
            index += 1
        while index < len(t):
            if t[index] not in letter_dict:
                letter_dict[t[index]] = -1
            else:
                letter_dict[t[index]] -= 1
            index += 1
        for val in letter_dict.values():
            if val != 0:
                return False
        return True
# @lc code=end

File: 242.有效的字母异位词.go
/*
 * @lc app=leetcode.cn id=242 lang=golang
 *
 * [242] 有效的字母异位词
 */

// @lc code=start
func isAnagram(s string, t string) bool {
    letterMap := map[byte]int{}
    index := 0
    m, n := len(s), len(t)
    for index < min(m, n) {
        letterMap[s[index]]++
        letterMap[t[index]]--
        index++
    }
    for index < m {
        letterMap[s[index]]++
        index++
    }
    for index < n {
        letterMap[t[index]]--
        index++
    }
    for _, val := range letterMap {
        if val != 0 {
            return false
        }
    }
    return true
}

func min(x, y int) int {
    if x < y {
        return x
    }
    return y
}
// @lc code=end

